packages/venia-concept/src/actions/cart/asyncActions.js

```diff
@@ -201,15 +199,15 @@ export const updateItemInCart = (payload = {}, targetItemId) => {
 
             // check if the guest cart has expired
             if (noGuestCartId || (response && response.status === 404)) {
-                // if so, then delete the cached ID...
-                // in contrast to the save, make sure storage deletion is
+                // Delete the cached ID from local storage and Redux.
+                // In contrast to the save, make sure storage deletion is
                 // complete before dispatching the error--you don't want an
                 // upstream action to try and reuse the known-bad ID.
-                await clearGuestCartId();
+                await dispatch(removeCart());
                 // then create a new one
                 await dispatch(createGuestCart());
-                // then retry this operation
-                return thunk(...arguments);
+                // and add the updated item to the new cart.
+                await dispatch(addItemToCart(payload));
```

Comment:
This may or may not be controversial so I'll point it out.

Here we've failed to update an item in the cart due to the cart having expired. We've just gotten a brand new (empty) cart, so there's no sense in retrying the update to this item because it won't exist. But we do know what the user wanted, so we add the updated item to the new cart.

Sucks that they lose anything else they had in the cart, but 🤷‍♂️

Reply:
Hmm, we'll have to think this one through.